Allow deferred datasets to behave as URIs

davelopez opened the pull request:
Closes #19065

Some tools might accept URIs as input. In those cases there is no need to materialize deferred datasets in Galaxy, we can just pass the source URI directly to the tool.

To do that, we can now use `allow_uri_if_protocol` when declaring the input param in the tool.

```xml
<param name="input" type="data" allow_uri_if_protocol="http,https,s3" />
```

With the above example, if you provide a deferred dataset as `input` which source URI starts with one of the listed protocols (http, https or s3), when you reference your `$input` it will contain the source URI, otherwise, it will be the path to the materialized file.

You can also use `allow_uri_if_protocol="*"` to treat all deferred datasets as URIs regardless of the protocol.

## How to test the changes?
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [x] Instructions for manual testing are as follows:
  - Upload a deferred dataset
  - Run a tool that uses `allow_uri_if_protocol` in the input param definition with an appropriate value to filter the protocol or `*`.
  - When you reference the input param in your command line it will contain the source URI of the deferred dataset.

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

davelopez committed:
Allow deferred datasets to behave as URI

nsoranzo commented on lib/galaxy/tool_util/xsd/galaxy.xsd:
```diff
@@ -4314,6 +4314,25 @@ template if the parameter is ``false`` or not checked by the user. Used only
+        <xs:attribute name="allow_uri_if_prefixed" type="xs:string">
+          <xs:annotation>
+            <xs:documentation xml:lang="en"><![CDATA[
+When using `deferred` datasets, Galaxy will try to materialize then into files when
```
```suggestion
When using `deferred` datasets, Galaxy will try to materialize them into files when
```

davelopez committed:
Fix typo

Co-authored-by: Nicola Soranzo <nicola.soranzo@gmail.com>

nsoranzo commented on lib/galaxy/tool_util/xsd/galaxy.xsd:
````diff
@@ -4314,6 +4314,25 @@ template if the parameter is ``false`` or not checked by the user. Used only
+This is useful when the tool can handle the URI directly. Example:
+
+```xml
+<param name="input" type="data" format="txt" allow_uri_if_prefixed="http|s3" />
````
Maybe we should be more specific/strict, and only allow exact protocols to be specified?

```suggestion
<param name="input" type="data" format="txt" allow_uri_if_protocol="http|https|s3" />
```

davelopez replied:
Sounds good to me :+1: Thanks!

bgruening replied:
Is there a reason to use | instead of a comma? For startups we also use commas

davelopez replied:
> Is there a reason to use | instead of a comma? For startups we also use commas

No reason beyond following the ideas in https://github.com/galaxyproject/galaxy/issues/19065, of course, we can use commas for consistency although I kind of like the `|` as lately there is more tendency to use it to describe unions like in [PEP604](https://peps.python.org/pep-0604/) or [Typescript](https://www.typescriptlang.org/docs/handbook/typescript-in-5-minutes-func.html#unions). 

That said, I'll change it to commas for consistency with the existing way :+1:

davelopez committed:
Rename allow_uri_if_prefixed to allow_uri_if_protocol

Co-authored-by: Nicola Soranzo <nicola.soranzo@earlham.ac.uk>

davelopez committed:
Use commas to list protocols in allow_uri_if_protocol

Co-authored-by: Bjoern Gruening <bjoern.gruening@gmail.com>

davelopez committed:
Add test for parsing allow_uri_if_protocol

davelopez committed:
Add `has_deferred_data` flag to MockDataset and FakeDatasetAssociation

davelopez committed:
Fix handling for missing deferred input in ToolEvaluator

davelopez committed:
Add test for allow_uri_if_protocol with deferred input

bgruening commented on lib/galaxy/tool_util/xsd/galaxy.xsd:
````diff
@@ -4314,6 +4314,25 @@ template if the parameter is ``false`` or not checked by the user. Used only
+This is useful when the tool can handle the URI directly. Example:
+
+```xml
+<param name="input" type="data" format="txt" allow_uri_if_protocol="http,https,s3" />
````
It would be nice to have a wildcard options here at some point.

davelopez replied:
Done in https://github.com/galaxyproject/galaxy/pull/19077/commits/18fb0a785e03e0463c92b056500d5525b4f229a2

bgruening commented on lib/galaxy/tools/evaluation.py:
```diff
@@ -327,6 +328,23 @@ def find_deferred_collections(input, value, context, prefixed_name=None, **kwarg
+        deferred_input = self.tool.inputs.get(input_name)
+        if not deferred_input:
+            # TODO: Can this ever happen? It seems like it happens in the test suite.
+            # For example, in test_metadata_validator_on_deferred_input
```
The BAM file in this test is a special case. BAM files do have a second file associated to it as metadata, a `bai` file.
The `bai` file is probably created during materialization. Can this trigger it here?

davelopez replied:
The inputs in the `pileup` tool are:
```xml
<inputs>
    <param name="input1" type="data" format="unsorted.bam" multiple="true" min="1" label="BAM Inputs">
        <validator check="bam_index" message="Metadata missing, click the pencil icon in the history item and use the auto-detect feature to correct this issue." type="metadata" />
    </param>
    <param name="reference" type="data" format="fasta" label="Fasta Input"/>
</inputs>
```
But during evaluation, we got 2 inputs (apart from `$reference`) as deferred `$input1` and  `$input11`. Maybe that `$input11` was the index file as you mentioned? I would have expected that to be under `$input1.metadata.bam_index` and not in its own input entry. Also at this point, nothing has been materialized yet, but maybe it was "staged" to be materialized... :thinking: 

Anyway, if this is expected I will just remove the TODO comment, I wanted to make sure this was a known case.

bgruening replied:
No idea what `$input11` is. @jmchilton can you please have a look? Is it ok to merge this PR or is this something we should look at?

davelopez replied:
> Is it ok to merge this PR or is this something we should look at?

If there is an issue (and that is a big IF) this PR is not related to it, it just "uncovered" it. So it should be safe to merge this if the review is positive :)

mvdbeek replied:
This is how multiple inputs are passed internally. I don't think you need to worry about this. It is not the bam index.

davelopez replied:
Thanks for the clarification!

bgruening reviewed (approved):
Great stuff @davelopez!

davelopez committed:
Enhance allow_uri_if_protocol to support wildcard '*' for deferred datasets

davelopez committed:
Remove redundant check for missing deferred input in ToolEvaluator

There might be cases with additional deferred inputs outside of the input definitions see test_metadata_validator_on_deferred_input.

bgruening reviewed (approved):
Looks good to me!

mvdbeek commented on test/functional/tools/parameters/gx_allow_uri_if_protocol.xml:
```diff
@@ -0,0 +1,13 @@
+<tool id="gx_allow_uri_if_protocol" name="gx_allow_uri_if_protocol" version="1.0.0">
+    <command><![CDATA[
+echo '$input' > '$output'
```
Realistically tools will have to support both materialized and deferred datasets, since you can pass both. Can you spec out how tool authors should handle that? Alternatively you could pass a file:// uri, but in either case we should document what a tool needs to handle.

davelopez replied:
Thanks for the good feedback!

I've expanded a bit the documentation in the XSD and the test tool. I hope this is what you had in mind, if not, let me know :+1:

mvdbeek commented on lib/galaxy_test/api/test_tools.py:
```diff
@@ -2573,6 +2573,24 @@ def test_metadata_validator_can_fail_on_deferred_input(self, history_id):
         assert job_details["state"] == "failed"
 
+    @skip_without_tool("gx_allow_uri_if_protocol")
+    def test_allow_uri_if_protocol_on_deferred_input(self, history_id):
```
I suppose it'll just work, but can you also test passing in a collection of deferred datasets here?

davelopez replied:
Done in https://github.com/galaxyproject/galaxy/pull/19077/commits/2bdcec0dda2d9e862fa7a043de6b0207223e43ca :)

davelopez committed:
Add is_deferred property to DatasetFilenameWrapper

For improved handling of deferred data

davelopez committed:
Expand documentation on allow_uri_if_protocol

davelopez committed:
Refactor deferred source URI handling in DatasetInstance and ToolEvaluator

davelopez committed:
Add test for gx_allow_uri_if_protocol with deferred collections
